File: src/completed/python/euler0004.py
###
#
# Created by Holden on 12/17/2015
#
# SOLVED on 12/17/2015
#
# Problem 4 - Largest Palindrome Product
#
###

def palindrome_product(digits):
    palindrome = 0
    
    for i in range((10 ** (digits - 1)), (10 ** digits)):
        for j in range((10 ** (digits - 1)), (10 ** digits)):
            if str(i * j) == (str(i * j))[::-1]:
                if i * j > palindrome:
                    palindrome = i * j
            
    return palindrome

File: java/completed/euler0004.java
import lib.EulerTime;

/**
 * Created by Holden on 11/19/2016. - COMPLETED
 *
 * Problem 4 - Largest Palindrome Product
 *
 */

public class Prob0004 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        int answer = largestPalindromeProduct(3);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static int largestPalindromeProduct(int integerLength) {
        int productMin = (int) Math.pow(10, integerLength - 1);
        int productMax = (int) (Math.pow(10, integerLength) - 1);
        int palindrome = 0;

        for (int factor1 = productMin; factor1 <= productMax; factor1++) {
            for (int factor2 = factor1; factor2 <= productMax; factor2++) {
                int product = factor1 * factor2;
                String productString = Integer.toString(product);
                String reverse = new StringBuilder(productString).reverse().toString();

                //System.out.print(productString);
                //System.out.print(new StringBuilder(productString).reverse().toString());

                if (productString.compareTo(reverse) == 0) {
                    if (product > palindrome) {
                        palindrome = product;
                    }
                }
            }
        }

        return palindrome;
    }
}